File: morris/rules.py
from __future__ import annotations

from typing import Iterable, List, Optional

from .constants import MILLS, ADJACENT, EMPTY, PLAYER_WHITE, PLAYER_BLACK


def positions_in_mill(board: List[int], idx: int) -> Optional[List[int]]:
	"""Return the mill triplet if idx is part of a formed mill for its occupant; else None."""
	player = board[idx]
	if player == EMPTY:
		return None
	for a, b, c in MILLS:
		if idx in (a, b, c) and board[a] == board[b] == board[c] == player:
			return [a, b, c]
	return None


def forms_mill(board: List[int], idx: int, player: int) -> bool:
	"""Check if placing/moving player at idx forms a mill."""
	for a, b, c in MILLS:
		if idx in (a, b, c):
			line = [a, b, c]
			if all((board[p] == player if p != idx else True) for p in line):
				# ensure all 3 equal to player, with idx being player's after move
				if all(board[p] == player for p in line if p != idx):
					return True
	return False


def legal_removals(board: List[int], opponent: int) -> List[int]:
	"""Return indices of opponent pieces that can be removed given a mill formed.

	Prefer removing pieces not in mills; only if all are in mills, allow removing any.
	"""
	opponent_positions = [i for i, v in enumerate(board) if v == opponent]
	if not opponent_positions:
		return []
	not_in_mill = [i for i in opponent_positions if positions_in_mill(board, i) is None]
	return not_in_mill if not_in_mill else opponent_positions


def neighbors_of(idx: int) -> List[int]:
	return ADJACENT[idx][:]

File: web/js/rules.js
import { MILLS, ADJACENT, EMPTY } from './constants.js';

export function positionsInMill(board, idx){
  const player = board[idx];
  if(player === EMPTY) return null;
  for(const [a,b,c] of MILLS){
    if((idx===a||idx===b||idx===c) && board[a]===player && board[b]===player && board[c]===player){
      return [a,b,c];
    }
  }
  return null;
}

export function formsMill(board, idx, player){
  for(const [a,b,c] of MILLS){
    if(idx===a || idx===b || idx===c){
      const line = [a,b,c];
      const othersOk = line.filter(p => p!==idx).every(p => board[p]===player);
      if(othersOk) return true;
    }
  }
  return false;
}

export function legalRemovals(board, opponent){
  const oppPositions = [];
  for(let i=0;i<board.length;i++) if(board[i]===opponent) oppPositions.push(i);
  if(oppPositions.length===0) return [];
  const notInMill = oppPositions.filter(i => positionsInMill(board, i)===null);
  return notInMill.length ? notInMill : oppPositions;
}

export function neighborsOf(idx){
  return [...ADJACENT[idx]];
}
